### PR TITLE
storage: don't serve requests on RHS after a merge commits

### LINKED ISSUE #30349: storage: TestStoreRangeMergeRHSLeaseExpiration failed under stress
SHA: https://github.com/cockroachdb/cockroach/commits/fd520679420038f598c97158672fb201d81ecc13

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeMergeRHSLeaseExpiration PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=911513&tab=buildLog

```
=== RUN   TestStoreRangeMergeRHSLeaseExpiration
I180918 06:58:50.486463 43084 storage/store.go:1559  [s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180918 06:58:50.494905 43084 gossip/gossip.go:408  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:37147" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I180918 06:58:50.495750 43371 gossip/gossip.go:940  [n1] gossip connectivity
  n1 [sentinel];
I180918 06:58:50.521979 43084 storage/store.go:1559  [s2] [n2,s2]: failed initial metrics computation: [n2,s2]: system config not yet available
W180918 06:58:50.535061 43084 gossip/gossip.go:1516  [n2] no incoming or outgoing connections
I180918 06:58:50.536012 43084 gossip/gossip.go:408  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:46009" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > build_tag:"" started_at:0 
I180918 06:58:50.537267 43218 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:37147
I180918 06:58:50.537700 43354 gossip/gossip.go:940  [n2] gossip connectivity
I180918 06:58:50.548579 43354 gossip/gossip.go:940  [n2] gossip connectivity
  n1 [sentinel];
  n2 -> n1;
I180918 06:58:50.549627 43371 gossip/gossip.go:940  [n1] gossip connectivity
  n1 [sentinel];
  n2 -> n1;
I180918 06:58:50.616152 43084 storage/replica_command.go:298  [s1,r1/1:/M{in-ax}] initiating a split of this range at key "b" [r2]
I180918 06:58:50.636569 43084 rpc/nodedialer/nodedialer.go:92  [s1,r1/1:{/Min-b}] connection to n2 established
I180918 06:58:50.641361 43084 storage/store_snapshot.go:615  [s1,r1/1:{/Min-b}] sending preemptive snapshot 8ec059d1 at applied index 18
I180918 06:58:50.642356 43084 storage/store_snapshot.go:657  [s1,r1/1:{/Min-b}] streamed snapshot to (n2,s2):?: kv pairs: 17, log entries: 8, rate-limit: 2.0 MiB/sec, 5ms
I180918 06:58:50.644031 42959 storage/replica_raftstorage.go:803  [s2,r1/?:{-}] applying preemptive snapshot at index 18 (id=8ec059d1, encoded size=2675, 1 rocksdb batches, 8 log entries)
I180918 06:58:50.647781 42959 storage/replica_raftstorage.go:809  [s2,r1/?:{/Min-b}] applied preemptive snapshot in 4ms [clear=0ms batch=0ms entries=2ms commit=1ms]
I180918 06:58:50.650181 43084 storage/replica_command.go:812  [s1,r1/1:{/Min-b}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:{/Min-b} [(n1,s1):1, next=2, gen=1]
I180918 06:58:50.656570 43084 storage/replica.go:3754  [s1,r1/1:{/Min-b},txn=aee937e7] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I180918 06:58:50.775692 43438 rpc/nodedialer/nodedialer.go:92  connection to n1 established
I180918 06:58:50.804595 43084 storage/store_snapshot.go:615  [s1,r2/1:{b-/Max}] sending preemptive snapshot a2a77f86 at applied index 12
I180918 06:58:50.805826 43084 storage/store_snapshot.go:657  [s1,r2/1:{b-/Max}] streamed snapshot to (n2,s2):?: kv pairs: 40, log entries: 2, rate-limit: 2.0 MiB/sec, 3ms
I180918 06:58:50.806867 43241 storage/replica_raftstorage.go:803  [s2,r2/?:{-}] applying preemptive snapshot at index 12 (id=a2a77f86, encoded size=7354, 1 rocksdb batches, 2 log entries)
I180918 06:58:50.808641 43241 storage/replica_raftstorage.go:809  [s2,r2/?:{b-/Max}] applied preemptive snapshot in 2ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I180918 06:58:50.810773 43084 storage/replica_command.go:812  [s1,r2/1:{b-/Max}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r2:{b-/Max} [(n1,s1):1, next=2, gen=0]
I180918 06:58:50.821680 43084 storage/replica.go:3754  [s1,r2/1:{b-/Max},txn=8fee3f2e] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I180918 06:58:50.864617 43388 storage/replica_proposal.go:211  [s2,r2/2:{b-/Max}] new range lease repl=(n2,s2):2 seq=2 start=0.000000123,248 epo=1 pro=0.000000123,249 following repl=(n1,s1):1 seq=1 start=0.000000000,0 exp=0.900000123,5 pro=0.000000123,7
I180918 06:58:50.925106 43494 storage/replica_command.go:430  [s1,r1/1:{/Min-b}] initiating a merge of r2:{b-/Max} [(n1,s1):1, (n2,s2):2, next=3, gen=0] into this range
I180918 06:58:50.950612 43084 storage/client_test.go:1243  test clock advanced to: 1.800000125,0
I180918 06:58:51.655080 43510 storage/client_merge_test.go:1190  starting get 0
I180918 06:58:51.656301 43512 storage/client_merge_test.go:1190  starting get 1
I180918 06:58:51.657517 43513 storage/client_merge_test.go:1190  starting get 2
I180918 06:58:51.658786 43514 storage/client_merge_test.go:1190  starting get 3
I180918 06:58:51.660030 43527 storage/client_merge_test.go:1190  starting get 4
I180918 06:58:51.661254 43528 storage/client_merge_test.go:1190  starting get 5
I180918 06:58:51.662523 43529 storage/client_merge_test.go:1190  starting get 6
I180918 06:58:51.664039 43530 storage/client_merge_test.go:1190  starting get 7
I180918 06:58:51.665673 43511 storage/node_liveness.go:756  [s1,r2/1:{b-/Max}] incremented n2 liveness epoch to 2
I180918 06:58:51.665984 43460 storage/client_merge_test.go:1190  starting get 8
I180918 06:58:51.666843 43496 storage/client_merge_test.go:1190  starting get 9
I180918 06:58:51.674127 43285 storage/replica_proposal.go:211  [s1,r2/1:{b-/Max}] new range lease repl=(n1,s1):1 seq=3 start=1.800000125,142 epo=1 pro=1.800000125,143 following repl=(n2,s2):2 seq=2 start=0.000000123,248 epo=1 pro=0.000000123,249
I180918 06:58:51.742695 43382 storage/store.go:2729  [s2,r1/2:{/Min-b},txn=e05709f5] removing replica r2/2
I180918 06:58:51.752045 43293 storage/store.go:2729  [s1,r1/1:{/Min-b}] removing replica r2/1
--- FAIL: TestStoreRangeMergeRHSLeaseExpiration (1.32s)
	client_merge_test.go:1219: expected RangeNotFound error from get during merge, but got <nil>

```

### PR DESCRIPTION
**Pull request by benesch**
Previously, it was possible for the leaseholder of the right-hand side
of a merge to erroneously serve requests after the merge committed. This
would occur when the RHS's watcher goroutine noticed that a merge
committed before the LHS on that store applied the merge commit trigger.
In this case, the watcher goroutine would unblock traffic, but that
traffic would be served normally because the RHS would not be marked as
destroyed. In effect, the watcher goroutine was assuming that the RHS
would always be marked as destroyed before it discovered that the merge
had committed, but there is no such guarantee.

The fix is to teach the watcher goroutine to mark the replica as
destroyed if the merge has committed before unblocking traffic.

With the fix in place, TestStoreRangeMergeRHSLeaseExpiration is no
longer flaky. For good measure, add a new test,
TestStoreRangeMergeWatcher, which more reliably reproduces the
consistency violation.

Fix #30349.

Release note: None

### CODE REVIEW

**Review by tbg — CHANGES_REQUESTED** [review]
Looks good to me, just some comments to make sure we're getting things right.

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-:-LO3YOZ7C2FWNoDVDFeL:b-u2fufg)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 2932 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-LO3XgQP-DDijctpNaVI:-LO3XgQP-DDijctpNaVJ:bar8cq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c709ca46879db4b0e499e8f5bbc042184fda6089/pkg/storage/replica.go#L2932)):*
> ```Go
> 			// definitely committed.
> 			mergeCommitted = true
> 		case roachpb.ABORTED:
> ```

Wow, lots of code here. I agree that it's all needed, but having these two paths is awkward (also makes me wonder whether it's really all tested), perhaps we can just always hit meta? I know there's a scaling concern here, but it might be one of those that remain forever theoretical. Also (and we have to test that stuff), is there some case where the meta range you're pinging here is the range itself and it just deadlocks?
___
*[pkg/storage/replica.go, line 2995 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-LO3Yg8r3D8LOXglxxn2:-LO3Yg8r3D8LOXglxxn3:b-i5t6dg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c709ca46879db4b0e499e8f5bbc042184fda6089/pkg/storage/replica.go#L2995)):*
> ```Go
> 		// error. If the merge aborted, the requests will be handled normally.
> 		r.mu.mergeComplete = nil
> 		close(mergeCompleteCh)
> ```

Everyone blocked on this checks the destroyStatus after this point, right?

**Review by benesch — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30986)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 2932 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-LO3XgQP-DDijctpNaVI:-LO3rPqA8nCQECN-ooCI:b-d3c2rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c709ca46879db4b0e499e8f5bbc042184fda6089/pkg/storage/replica.go#L2932)):*
Coverage is actually quite good. I made it just a little bit better by injecting failures into the push txn and get requests. Here's the coverage report:

https://rawgit.com/benesch/48af9b89aa4f588cc4ead5e92c7f7585/raw/0c3a33ba7500a5d55b9add6546e315dfbd044753/coverage.html#file27

I'm going to add a test case shortly to hit one last branch that isn't covered.

> perhaps we can just always hit meta? I know there's a scaling concern here, but it might be one of those that remain forever theoretical.

That introduces its own complexities, though. Ideally, we'd just do a consistent read from meta2. That would both wait for the merge txn to complete, by waiting for its intent to get resolved, and tell you how that intent was resolved! But I tried that, and you can end up deadlocking on yourself. See https://github.com/cockroachdb/cockroach/blob/69ce79fe832e5a26b577b50b4310909e070bb482/pkg/storage/client_merge_test.go#L995-L1032.

So, ok, you can poll meta2 with at the READ_UNCOMMITTED iso level instead. But how do you know when to stop polling? You can't just wait for the intent to disappear, because the intent might never disappear if the coordinator dies immediately after laying down the meta2 intent. The txn record will get GC'd without knowledge of the meta2 intent, and nothing will ever come looking for it. So you'd need to do some sort of intent resolution yourself. And I think that's at least as complex, if not more complex, than this solution.

> Also (and we have to test that stuff), is there some case where the meta range you're pinging here is the range itself and it just deadlocks?

Yep. The plan is to just disallow meta2 merges. I know we're getting down to the wire, but this is in my queue. https://github.com/cockroachdb/cockroach/issues/28886
___
*[pkg/storage/replica.go, line 2995 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-LO3Yg8r3D8LOXglxxn2:-LO3rPFQ6bFCvrzLs-eG:bb77uci) ([raw file](https://github.com/cockroachdb/cockroach/blob/c709ca46879db4b0e499e8f5bbc042184fda6089/pkg/storage/replica.go#L2995)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Everyone blocked on this checks the destroyStatus after this point, right?
</blockquote></details>

Yep!

**Review by tbg — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-:-LO3w6VV7vbKWA_xiB_x:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica.go, line 2932 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-LO3XgQP-DDijctpNaVI:-LO3voNb4DTvnWvL8RwI:b6rgr0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/c709ca46879db4b0e499e8f5bbc042184fda6089/pkg/storage/replica.go#L2932)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Coverage is actually quite good. I made it just a little bit better by injecting failures into the push txn and get requests. Here's the coverage report:

https://rawgit.com/benesch/48af9b89aa4f588cc4ead5e92c7f7585/raw/0c3a33ba7500a5d55b9add6546e315dfbd044753/coverage.html#file27

I'm going to add a test case shortly to hit one last branch that isn't covered.

> perhaps we can just always hit meta? I know there's a scaling concern here, but it might be one of those that remain forever theoretical.

That introduces its own complexities, though. Ideally, we'd just do a consistent read from meta2. That would both wait for the merge txn to complete, by waiting for its intent to get resolved, and tell you how that intent was resolved! But I tried that, and you can end up deadlocking on yourself. See https://github.com/cockroachdb/cockroach/blob/69ce79fe832e5a26b577b50b4310909e070bb482/pkg/storage/client_merge_test.go#L995-L1032.

So, ok, you can poll meta2 with at the READ_UNCOMMITTED iso level instead. But how do you know when to stop polling? You can't just wait for the intent to disappear, because the intent might never disappear if the coordinator dies immediately after laying down the meta2 intent. The txn record will get GC'd without knowledge of the meta2 intent, and nothing will ever come looking for it. So you'd need to do some sort of intent resolution yourself. And I think that's at least as complex, if not more complex, than this solution.

> Also (and we have to test that stuff), is there some case where the meta range you're pinging here is the range itself and it just deadlocks?

Yep. The plan is to just disallow meta2 merges. I know we're getting down to the wire, but this is in my queue. https://github.com/cockroachdb/cockroach/issues/28886
</blockquote></details>

Thanks for the additional color, that looks good. Carry on. Disallowing m2 merges sounds fine for 2.1. Doubt we'll need them for quite some time to come (though I hope we can iron them out while you're still in this area).

**Comment by tschottdorf** [comment]
Oh - one more request - can you stress the crap out of the new/altered test before you merge?

**Review by benesch — COMMENTED** [review]
I added a new test case for the one branch I hadn't covered... and it caught a segfault. Whoops. The test is rather involved—would you mind taking a quick look? 

Stress is looking good after a few minutes. I'll let it run for 30+ before I merge. (I'll also wait for your re-review, of course.)

> Dismissed @tschottdorf from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-:-LO5-xem6-HkpoV9Rluc:b3ht33j)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/client_merge_test.go, line 2210 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-LO5-bHyBk_HvaSDRNGD:-LO5-bHyBk_HvaSDRNGE:b-x97vd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/06255e58d6de5d3c68447779ed2ee6a695085448/pkg/storage/client_merge_test.go#L2210)):*
> ```Go
> 	if atomic.LoadInt64(&sawMeta2Req) != 1 {
> 		t.Fatalf("test did not generate expected meta2 get request/response")
> 	}
> ```

This test is long and involved, but it caught a segfault, so I think it's worth keeping.
___
*[pkg/storage/replica.go, line 2966 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-LO5-d3S9HLUDFURKHfl:-LO5-d3S9HLUDFURKHfm:b-rekf37) ([raw file](https://github.com/cockroachdb/cockroach/blob/827907d357936239bef5491fb91668542771960a/pkg/storage/replica.go#L2966)):*
> ```Go
> 				break
> 			}
> 			if getRes.Value.RawBytes == nil {
> ```

When a Get request does not find a value, `getRes.Value` is nil. Not `getRes.Value.RawBytes`, as the comment on `roachpb.GetResponse` would have you believe. I've fixed the comment.

**Review by tbg — APPROVED** [review]
:lgtm: very happy about this test coverage.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-:-LO7d821790HcSf5xTSx:b-jjm3is)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/client_merge_test.go, line 2080 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-LO7byZU3510qH_I6sJn:-LO7byZVD5M8rJs4kjbZ:b-81eq6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/06255e58d6de5d3c68447779ed2ee6a695085448/pkg/storage/client_merge_test.go#L2080)):*
> ```Go
> // merge occurs.
> //
> // This test is a more complicated version of TestSToreRangeMergeWatcher that
> ```

STore

**Review by benesch — COMMENTED** [review]
TFTR!

bors r=tschottdorf

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-:-LOGsuG31tdt0wOTFkBT:b5yfumx)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/client_merge_test.go, line 2080 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-LO7byZU3510qH_I6sJn:-LOGsYE3BgWy6UsP-SV_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/06255e58d6de5d3c68447779ed2ee6a695085448/pkg/storage/client_merge_test.go#L2080)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

STore
</blockquote></details>

Done.
___
*[pkg/storage/client_merge_test.go, line 2100 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/30986#-LOGscEk0YfcVjUmdCQU:-LOGscEk0YfcVjUmdCQV:b-hsvxf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/04345777caf7658b974a43e71ee3cfee4aedd492/pkg/storage/client_merge_test.go#L2100)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	t.Skip("flawed test: deadlocks if merge transaction retries")
> ```

Naturally, this test is great at timing out under race. Equally naturally, our PR stresser doesn't catch it because its timeout is 25m and the PR stresser only runs stress for 5m. Sigh. I don't have time to fix it right now but I really want to get this into the next v2.1 release, so I'm just going to merge with the test skipped.

**Pull request merged**
